README.md
# NAME

Catmandu::Store::CHI - a CHI backed caching store

# SYNOPSIS

    # From the command line
    $ catmandu export CHI --driver File --root_dir /data to YAML
    $ catmandu import JSON to CHI --driver File --root_dir /data < data.json

    # From perl
    use Catmandu;

    my $store = Catmandu->store('CHI', driver => 'File' , root_dir => '/data');

    $store->bag->each(sub {
         my $item = shift;
         ...
    });

    $store->bag->add({ test => 123 });

# METHODS

## new()

## new(driver => $chi\_driver, \[OPT => VAL, OPT2 => VAL\])

Create a new Catmandu::Store::CHI with a $chi\_driver and optional parameters. When no driver is given
then by default the 'Memory' driver will be used. See [CHI](https://metacpan.org/pod/CHI) for more documentation on possible drivers.

# INHERITED METHODS

This Catmandu::Store implements:

- [Catmandu::Store](https://metacpan.org/pod/Catmandu::Store)

Each Catmandu::Bag in this Catmandu::Store implements:

- [Catmandu::Bag](https://metacpan.org/pod/Catmandu::Bag)

# SEE ALSO

[CHI](https://metacpan.org/pod/CHI)

# AUTHOR

Patrick Hochstenbach, `<patrick.hochstenbach at ugent.be>`

# LICENSE AND COPYRIGHT

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for Catmandu-Store-CHI

{{$NEXT}}

0.04  2019-10-10 20:37:45 CEST
  - fixing dependencies
  
0.03  2019-10-10 16:33:52 CEST
  - removing Data::UUID dependency
  - cosmetic changes

0.02  2014-06-28 08:50:41 CEST
  - adding pod
  - adding Dist::Milla
  - adding tests

0.01    2014-06-28
  - initial release


test.pl
#!/usr/bin/env perl

use lib qw(lib);
use Catmandu;
use Data::Dumper;

my $store = Catmandu->store('CHI');

$store->add({test => 123});
$store->add({test => 645});

$store->each(sub {
	my $obj = shift;
	print Dumper($obj);
});

Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.008.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "a CHI backed caching store",
  "dist_author" => [
    "Patrick Hochstenbach, C<< <patrick.hochstenbach at ugent.be> >>"
  ],
  "dist_name" => "Catmandu-Store-CHI",
  "dist_version" => "0.03",
  "license" => "perl",
  "module_name" => "Catmandu::Store::CHI",
  "recursive_test_files" => 1,
  "requires" => {
    "CHI" => "0.58",
    "Catmandu" => "1.2001",
    "Moo" => "1.0",
    "perl" => "v5.10.1"
  },
  "test_requires" => {
    "Test::More" => "1.001003",
    "Test::Simple" => "1.001003"
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.28",
  "Test::More" => "1.001003",
  "Test::Simple" => "1.001003"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;


lib/Catmandu/Store/CHI/Bag.pm
package Catmandu::Store::CHI::Bag;

use Moo;
use CHI;
use Data::UUID;
use Data::Dumper;

with 'Catmandu::Bag';

has 'chi'    => (is => 'ro' , lazy => 1 , builder => 1);

sub _build_chi {
    my ($self) = @_;
    my $driver = $self->store->driver;
    my $opts   = $self->store->opts;
    my $name   = __PACKAGE__ . '::' . $self->name;
    CHI->new(namespace => $name , driver => $driver, %$opts);
}

sub generator {
    my $self = shift;
    my $keys = [ $self->chi->get_keys ];
    sub {
        my $id = shift @$keys;

        return undef unless $id;

        $self->get($id);
    };
}

sub get {
    my ($self,$id) = @_;
    $self->chi->get($id);
}

sub add {
    my ($self,$data) = @_;
    $data->{_id} //= gen_id();
    my $id = $data->{_id};
    $self->chi->set($id,$data);
    return $data;
}

sub delete {
    my ($self,$id) = @_;
    $self->chi->remove($id);
}

sub delete_all {
    my ($self) = @_;
    $self->chi->clear();
}

sub gen_id {
    Data::UUID->new->create_str();
}

1;

lib/Catmandu/Store/CHI.pm
=head1 NAME

Catmandu::Store::CHI - a CHI backed caching store

=head1 SYNOPSIS

   # From the command line
   $ catmandu export CHI --driver File --root_dir /data to YAML
   $ catmandu import JSON to CHI --driver File --root_dir /data < data.json

   # From perl
   use Catmandu;

   my $store = Catmandu->store('CHI', driver => 'File' , root_dir => '/data');

   $store->bag->each(sub {
        my $item = shift;
        ...
   });

   $store->bag->add({ test => 123 });

=head1 METHODS

=head2 new()

=head2 new(driver => $chi_driver, [OPT => VAL, OPT2 => VAL])

Create a new Catmandu::Store::CHI with a $chi_driver and optional parameters. When no driver is given
then by default the 'Memory' driver will be used. See L<CHI> for more documentation on possible drivers.

=head1 INHERITED METHODS

This Catmandu::Store implements:

=over 3

=item L<Catmandu::Store>

=back

Each Catmandu::Bag in this Catmandu::Store implements:

=over 3

=item L<Catmandu::Bag>

=back

=head1 SEE ALSO

L<CHI>

=head1 AUTHOR

Patrick Hochstenbach, C<< <patrick.hochstenbach at ugent.be> >>

=head1 LICENSE AND COPYRIGHT

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut
package Catmandu::Store::CHI;

{
  $Catmandu::Store::CHI::VERSION = '0.03';
}

use Moo;
use CHI;
use Catmandu::Store::CHI::Bag;

with 'Catmandu::Store';

has 'driver' => (is => 'ro' , required => 1 , default => sub { 'Memory' });
has 'opts'   => (is => 'rw');

sub BUILD {
    my ($self,$opts) = @_;

    if (keys %$opts == 0) {
        $opts->{global} = 1;
    }

    delete $opts->{driver};

    $self->opts($opts);
}

1;
